[process-agent] Chunk payloads (CollectorProc) by desired count and weight

xornivore opened the pull request:
### What does this PR do?

- Replace existing chunking with chunking taking into account size of the payload derived from properties such as CLI/username.
- Allow spliting processes for a single container across multiple payloads (by duplicating the container).
- Deprecate `process_config.max_ctr_procs_per_message` setting (no longer used, `process_config.max_per_message` is used for size limit in all cases).
- Code is generic enough to be reused for other payload types (see tests using `testPayload`).
- Add `process_config.max_message_bytes` with default of 1MB (before compression) for the target cumulative size of process payloads.

### Motivation

Avoid situations when the payload becomes exceedingly large.

### Possible Drawbacks / Trade-offs

- Previous chunking logic is simpler in a sense that it can take a slice of payloads without inspecting, we need to profile the contribution of this changes to overall performance of the process check.
- Intentionally not using protobuf `Size` - while it's possibly accurate it would result in checking every field.

### Describe how to test/QA your changes

- Test on an OS having a mix of processes running within containers. Validate that processes and containers are reported correctly.
- Test with processes having very large CLIs (~1MB). Validate that they are reported correctly and we are not missing any data for other processes.
- For testing on Kubernetes you can use this tool: [xornivore/biggie](https://github.com/xornivore/biggie), it allows creating arbitrary number of processes with arbitrarily long command lines:
  - `ARG_LEN` controls length of a single CLI argument.
  - `ARG_COUNT` controls count of arguments
  - `PROC_COUNT` controls the number of processes to create.
  See [biggie-deployment.yaml](https://github.com/xornivore/biggie/blob/main/biggie-deployment.yaml) for configuration.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

Commit:
[process-agent] Chunk process payloads (CollectorProc) by desired count and weight

- Replace existing chunking with chunking taking into account size of the payload derived from properties such as CLI/username.
- Allow spliting processes for a single container across multiple payloads (by duplicating the container).

djmitche reviewed (approved):
Config change looks good for agent-core

maycmlee commented on releasenotes/notes/process-agent-weight-aware-message-chunking-b534a02b06123d53.yaml:
```diff
@@ -0,0 +1,9 @@
+enhancements:
+  - |
+    Changes process payload chunking in the process Agent to take into account
+    size of process details such as CLI, and user name.
```
```suggestion
    the size of process details such as CLI and user name.
```

maycmlee reviewed (commented):
Small edits

Commit:
Update releasenotes/notes/process-agent-weight-aware-message-chunking-b534a02b06123d53.yaml

Co-authored-by: May Lee <may.lee@datadoghq.com>

just-chillin commented on pkg/process/config/config.go:
```diff
@@ -95,19 +95,18 @@ type WindowsConfig struct {
-	Blacklist                 []*regexp.Regexp
-	Scrubber                  *DataScrubber
-	MaxPerMessage             int
-	MaxCtrProcessesPerMessage int // The maximum number of processes that belong to a container for a given message
```
Noice 👍🏻

xornivore replied:
It's not bloat, it's code, someone's hard work 😄

just-chillin reviewed (approved):
Looks great! This will be a huge improvement. Just a few documentation specific points

mbotarro commented on pkg/process/checks/chunking.go:
```diff
@@ -0,0 +1,168 @@
+	model "github.com/DataDog/agent-payload/v5/process"
+)
+
+type payloadList interface {
```
nit: what do you think about giving a brief description of each interface, its methods and how it's intended to be used?

xornivore replied:
Yes, good idea, I think it's not obvious.

mbotarro commented on pkg/process/checks/chunking.go:
```diff
@@ -0,0 +1,168 @@
+	idx   int
+}
+
+func (c *chunkPropsTracker) TakenSize() int {
```
nit: what do you think of adding comments to make it explicit that `processList` implements the `payloadList` interface and `chunckPropsTracker` implements the `chunkAllocator`?
It wasn't super clear to me when I first saw the code.

wiyu replied:
We should add the compile time interface check to achieve the above. 
```
var _ chunkAllocator = & chunkPropsTracker{}
```

xornivore replied:
that's a great idea thank you @wiyu @mbotarro

mbotarro commented on pkg/process/checks/chunking.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+type processChunker interface {
+	Accept(procs []*model.Process, weight int)
```
nit: maybe add a comment saying that Accept is what is actually building each chunk?

xornivore replied:
Will do.

mbotarro commented on pkg/process/checks/chunking.go:
```diff
@@ -0,0 +1,168 @@
+	l.chunker.Accept(l.procs[start:end], weight)
+}
+
+type chunkProps struct {
+	weight int
+	size   int
+}
+type chunkPropsTracker struct {
+	props []chunkProps
+	idx   int
+}
```
what do you think about calling these ?
```golang
type chunk struct {
	weight int
	size   int
}
type chunkTracker struct {
	chunks []chunk
	idx   int
}
```
`weight` and `size` are properties from a `chunk` so I'm not sure if we need to call the struct `chunkProps`. What do you think?

mbotarro commented on pkg/process/checks/chunking_test.go:
```diff
@@ -0,0 +1,622 @@
+						// chop!
+						1,
+						1,
+						1,
```
is it a `// chop!` comment missing here?

xornivore replied:
Hm, nice catch let me investigate!

mbotarro commented on pkg/process/checks/process.go:
```diff
@@ -255,116 +255,56 @@ func createProcCtrMessages(
-			space = capacity
+		totalProcs += len(procs)
+		totalContainers++
+		chunker.container = ctr
+		chunker.idx = 0
+		if len(chunker.collectorProcs) > 1 {
+			chunker.idx = len(chunker.collectorProcs) - 1
```
I wonder if we could encapsulate the `idx` initialization inside of the chunker. And maybe instead of setting
```
chunker.container = ctr
```
we could pass the container as a parameter in the `chunkProcessesBySizeAndWeight` function and set the chunker container inside of it.

xornivore replied:
This is an interesting idea, let me see if I can make it work.

xornivore commented on pkg/process/checks/chunking.go:
```diff
@@ -0,0 +1,168 @@
+	Next()
+}
+
+func chunkPayloadsBySizeAndWeight(l payloadList, a chunkAllocator, maxChunkSize int, maxChunkWeight int) {
```
Makes sense, I will add that.

xornivore commented on pkg/process/checks/process_nix_test.go:
```diff
@@ -3,9 +3,6 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build linux
```
Good point, I haven't renamed because I suspect it may create conflict with other PRs in flight, I will do it after however.

xornivore commented on releasenotes/notes/process-agent-weight-aware-message-chunking-b534a02b06123d53.yaml:
```diff
@@ -0,0 +1,9 @@
+    the size of process details such as CLI and user name.
+deprecations:
+  - |
+    Deprecates process_config.max_ctr_procs_per_message setting.
```
Yes, this is more accurate thank you!

Commit:
Address review feedback

Commit:
Improve weighProcess implementation to be closer to proto serialization

Commit:
Fix comment in test

just-chillin reviewed (approved):
Looks much better, just want to make sure that @mbotarro is aware that this PR also removes `max_ctr_procs_per_message`

xornivore commented on releasenotes/notes/process-agent-weight-aware-message-chunking-b534a02b06123d53.yaml:
```diff
@@ -0,0 +1,9 @@
+    the size of process details such as CLI and user name.
+deprecations:
+  - |
+    Removes the process_config.max_ctr_procs_per_message setting.
```
Yes, he is changing this setting, I will rebase

Commit:
Add process_config.max_message_bytes setting

Additionally refactors the validation helpers in pkg/process/checks:
- sync.Once is overkill for this - settings are already loaded in memory
- simplify testing
- once we have New<CheckType> initializers we can move some of these validators to be in `cmd/process-agent`
